### ISSUE TITLE
Testing results for 2D Unified Batching (GLES2 & GLES3)

### ISSUE THREAD

**Issue by lawnjelly** [godotengine/godot#42899]
## Important
`rendering->batching->options->use_batching_in_editor` needs to be set to true for 3.2.4 beta 1.
Sorry, this is due to a small bug on my part. Otherwise you will get a multicolored IDE! :grinning: 

## About
This issue is to collect user feedback about the new 2d unified batching for 3.2.4 implemented in #42119. This PR affects GLES2 users as well as GLES3, there are several new improvements to GLES2 batching, including new primitives, improved support for batching custom shaders, and software skinning.

* Batching is enabled by default in both games and the editor
* Batching is totally optional, it can be switched off in project settings
* The project settings can nearly all be found in `rendering->batching`

To some extent this post mirrors information in #38004 as the unified batching is largely based on the GLES2 batching codebase.

See for details as to how batching works:
https://docs.godotengine.org/en/stable/tutorials/optimization/batching.html

## Introduction
3.2.4 introduces unified 2D batching support to GLES2 & GLES3. This is quite a major modification to the renderer, so we are expecting a certain amount of issues, it would be helpful to keep them here together so we can work through them.

In the old renderer, 2d primitives were drawn individually. This is simple to implement, but doesn't make best use of the speed of modern GPUs, which work far more efficiently when you tell them to draw a bunch of similar primitives at once. Batching works by placing an intermediate layer between Godot and the backend renderer, which groups similar primitives together so that they can be drawn at once. And this intermediate layer is now shared between backends, so that the same improvements that were made to GLES2 are now available to GLES3.

Note that batching is purely a performance optimization. If working correctly there should be no visual differences or differences in behaviour. It can work on several primitives including rects, lines and polys.

## Expectations
Some games will benefit from batching more than others. In particular games using tilemaps or a large amount of text are most likely to benefit.

**If you notice a slowdown** in your game, it is highly likely to be because batching defaults to a slower method of drawing, which prevents flicker on nvidia hardware. You can switch this back to the old behaviour by setting `rendering->batching->options->single_rect_fallback`.

**If you are using 2d lights** you may also find it useful to tweak `rendering->batching->lights->scissor_area_threshold`. You can get significant improvement to speed using 2d lights in some circumstances.

## Mac Users
If you experience performance problems, you are highly encouraged to try the experimental settings to change how OpenGL is used, and report back on your findings. Full details here:
https://github.com/godotengine/godot/issues/43038#issuecomment-722336615

## Additional
Along with performance improvements due to batching, the PR also introduces optional software skinning for 2d skeletal animations. This should hopefully greatly improve compatibility of skinning, especially on mobile phones, and may lead to speed increases. See `rendering->quality->2d->use_software_skinning`.

There is also the option to select between the current and old ninepatch rect behaviour, see `rendering->quality->2d->ninepatch_mode`.

## Bugs
There are 2 types of issue you are most likely to see:
### 1) Performance issues
We are particularly interested if you have a project that runs slower with batching than without (and is not cured by turning on `single_rect_fallback`, see above). If we can isolate these cases and work out what is going on, we can fix them to prevent any slowdowns in the next version (and in some cases they can be accelerated).

If you get an increase in performance, let us know! :+1: 

### 2) Visual regressions
If you encounter visual differences between your project with and without batching, that is something we need to fix. Please let us know full details.

## A Quick Guide
Settings are described in:
https://docs.godotengine.org/en/stable/tutorials/optimization/batching.html#project-settings

## Tips
The order that objects are rendered generally follows the order in the scene tree. If you can group similar items together in the scene tree (similar material, texture etc), they are more likely to be batched together, and give you a speed increase.

Batching will be broken by (among other things):
* Texture changes (not just changes in texture coordinates)
* Material changes
* Changes other than rect position, transform, UV coordinates and color

The best batching project settings will depend upon the game. Be sure to try varying every one, to see the effect on performance. A simple way to measure FPS is to disable vsync, and set the project setting `debug/settings/stdout/print_fps` (although you should disable this for exports, as printing anything can affect performance itself).

### Diagnose Frame
This is a very useful option in order to optimize your game (make sure to switch it off when exporting though, as it costs performance). It prints out every 10 seconds a list of the batches used in the frame. This list may be quite long for complex games, you may need to increase the buffer size in `project_settings/network/limits/debugger_stdout/max_chars_per_second` to read it all.

Details are in the documentation.
https://docs.godotengine.org/en/stable/tutorials/optimization/batching.html#diagnostics

### Template
As with most issues, it is useful if you provide us with the following info:
* Godot version
* OS including version, device, and GPU (if you know it)
* Issue description
* Steps to reproduce
* Minimum reproduction project

In addition if you are dealing with a slowdown, it can be helpful to provide (or link to) your `diagnose_frame` log.

If you are reporting increase in performance, only minimal info is needed (is totally up to you). :+1:

**Comment by fire** [comment]
How does one keep 2d batching?

In https://github.com/fire/Godot-Control-Masking we were afraid it would cause batching to fail, since the mask method runs the gpu shader on all the children. Even if masking was an engine module it would be effected.

**Comment by lawnjelly** [comment]
It is hard to say in advance, but if you run with `diagnose_frame` you will see whether they are batching. If it cannot, it will revert to the old method of drawing.

**Comment by oeleo1** [comment]
All this sounds terrific. At first glance, it looks terrific too. :-)

![image](https://user-images.githubusercontent.com/40200623/96799297-1e418580-1403-11eb-81e5-ae139ed99b10.png)

The `[redering]` section of the `project.godot` has these among others:
    ...
    quality/driver/driver_name="GLES2"
    ...
    batching/options/use_batching=false
    batching/options/use_batching_in_editor=false

If I quit, remove (manually) the line about `use_batching_in_editor=false` and restart, the UI looks normal.

At startup (Windows 10, 64-bit) it says invariably that batching is on though:
```
Godot Engine v3.2.4.beta1.official - https://godotengine.org
OpenGL ES 2.0 Renderer: Quadro T2000 with Max-Q Design/PCIe/SSE2
OpenGL ES Batching: ON
```
At your service to help debugging it.

**Comment by bitmapman5** [comment]
Just wanted to pop in and quickly say thanks for this! I've been working on a dwarf fortress-like colony sim in Godot 3.2.3 stable and had resigned myself to using GLES2 for the batching. With this PR it looks like rendering the massive pile of tilemaps in my game is in the neighborhood of 6x faster on GLES3, about the same speed as GLES2 on 3.2.3. 
I haven't noticed any issues so far but I'll come back and make a proper report if I do. 
Thank you!

**Comment by Oskit-Producciones** [comment]
GLES3:
Batching: off
![image](https://user-images.githubusercontent.com/663998/96816877-7cb33780-13f4-11eb-830f-c95ed41b271c.png)

Batching: on
![image](https://user-images.githubusercontent.com/663998/96816972-b84e0180-13f4-11eb-8022-581cfd0771a4.png)

**Comment by Oskit-Producciones** [comment]
single_rect_fallback: on
![image](https://user-images.githubusercontent.com/663998/96819952-41683700-13fb-11eb-97b5-d7016d1e3d91.png)

single_rect_fallback: off
![image](https://user-images.githubusercontent.com/663998/96820010-60ff5f80-13fb-11eb-8f11-cc130174daa8.png)

**Comment by lawnjelly** [comment]
@Oskit-Producciones That sounds good.

The drop in performance for single_rects is totally expected. The 'safe' routine is 2-3x slower. However on balance in most projects there will be a lot of batched primitives, which will more than make up for this. And you can always turn the `single_rect_fallback` on (if you don't mind some of your uses having flicker).

@oeleo1 I'll try and work out what is happening for you. Good news is though that 'big regressions' tend to be as easy or easier to fix than small regressions (strangely enough).

**Comment by avencherus** [comment]
> when I set batching to false and use_batching_in_editor to false!

That's a pretty wild one.  I feel for you here.  😆

Can also confirm **Windows 10 - OpenGL ES 2.0 Renderer: GeForce GTX 1660 SUPER/PCIe/SSE2**

**Comment by bones527** [comment]
[metroidvania.zip](https://github.com/godotengine/godot/files/5424599/metroidvania.zip)

Godot Version: 3.2.4beta1
OS: Endeavour OS
Device: HP 2000 Notebook
GPU: Intel HD 4000

After adding a wind shader to the tilemap, the tilemap disappears randomly at random points. The project file is included.

Steps to reproduce:

1. Open the project
2. Run the project
3. You can't see the grass without moving to the top platform

**Comment by lawnjelly** [comment]
> [metroidvania.zip](https://github.com/godotengine/godot/files/5424599/metroidvania.zip)
> 

Yes I can replicate this here. I've been adding some more code in 3.2.4 to deal with more custom shaders, so I expected there to be more regressions to fix here before everything is running smoothly.

The diagnose log shows it seems to be splitting it up correctly, so that is good.

**Comment by lawnjelly** [comment]
> [metroidvania.zip](https://github.com/godotengine/godot/files/5424599/metroidvania.zip)

Ok I am back home and able to compile debug etc and have this fixed now. :+1: 

As I suspected it was doing the transform twice (it uses hardware transform mode under some circumstances, such as only 1 rect in a batch), and it also now stores the parent transform in the new large flexible vertex format for dealing with custom shaders.

You may find that if you used the same shader on a batch containing several rects with the same texture, it would work (because hardware transform not used). That's probably how I initially tested it so didn't notice this problem.

The solution was to disallow hardware transform under all circumstances when using large_fvf. I still want to go over it tommorrow when I've had a nights sleep before making a PR, just to check there is no side effects, or a better way.

This is encouraging that the custom shader batching seems to be working. That was one of the major weak points of the 3.2.2 batching. I'm hoping most custom shaders should be batched now so even advanced games will benefit.

**Comment by QbieShay** [comment]
I get a crash when opening a certain scene. I'm using gles3, batching and batching in editor are on.

Here is the repro:
[repro-batch-gles3.zip.zip](https://github.com/godotengine/godot/files/5451734/repro-batch-gles3.zip.zip)

Navigate to effects/enemies/projectile/InfectionProjectile/BulletEffectInfection.tscn. Open the scene, editor crashes with SIGILL (4)

PC specs:
OS: Linux 5.9.1-arch1-1
GPU: Radeon 590, driver amdgpu

**Comment by akien-mga** [comment]
@QbieShay Thanks! Might be the same bug as #42994 (fix merged a few hours ago) as I can't reproduce the crash in latest `3.2` (efdc5f0f899a8313cd557f87cc7d478692ebdb72 + some cherry-picks I haven't pushed yet).

**Comment by QbieShay** [comment]
@lawnjelly @akien-mga it isn't happening anymore on `0b6d`

**Comment by lawnjelly** [comment]
> @lawnjelly @akien-mga it isn't happening anymore on `0b6d`

Yes I think the fix for this was #43102. There's just the faintest possibility of a similar problem with polys but I've not been able to produce it and had no reports (it might be that polys are handled fine by the existing logic).

**Comment by EIREXE** [comment]
Trying latest 3.2 (adf2dabbd) I get some strange artifacting around some of my graphics, they are controls being scaled in and out within a viewport rendered on top of a meshinstance.

This happens with UV contract and on windows and linux.

Big performance jump though, particularly from character drawing.

https://youtu.be/tUUygHPq2yI

**Comment by lawnjelly** [comment]
> Trying latest 3.2 ([adf2dab](https://github.com/godotengine/godot/commit/adf2dabbde23fdd89803871e26dd1d33ca20bf77)) I get some strange artifacting around some of my graphics, they are controls being scaled in and out within a viewport rendered on top of a meshinstance.
> 
> This happens with UV contract and on windows and linux.

If you are able to make a minimum reproduction project, I can investigate.

**Comment by EIREXE** [comment]
> 
> 
> > Trying latest 3.2 ([adf2dab](https://github.com/godotengine/godot/commit/adf2dabbde23fdd89803871e26dd1d33ca20bf77)) I get some strange artifacting around some of my graphics, they are controls being scaled in and out within a viewport rendered on top of a meshinstance.
> > This happens with UV contract and on windows and linux.
> 
> If you are able to make a minimum reproduction project, I can investigate. It does look like a filtering / precision issue.

Hmm I think it's gonna be hard but i'll give it a go, also I found some strange geometry that's drawn by the game (but doesn't seem to be visible) with batching enabled, not sure what that is or if it may even be my fault.

![imagen](https://user-images.githubusercontent.com/3101690/98914396-e93bc680-24c8-11eb-98a7-addba45c48e1.png)

(left is batching on, right is batching off)

**Comment by EIREXE** [comment]
I've tried reproducing the issue myself by isolating menu assets on a new project and I can't seem to make it happen on an isolated project, only inside my game, which is quite odd.

I also tried disabling some nodes and it appears like that changes the artifacting quite a bit:

https://youtu.be/ZXRjHJuFbxU

**Comment by lawnjelly** [comment]
Sometimes it can be difficult to deduce what is needed for a minimum reproduction project, but that is really the first step that is needed, especially in game project with a lot going on, so I would encourage you to keep at it. If you can't reproduce it, it is far less likely I will be able to, without knowing how your scene is arranged etc.

I would do exactly as you are doing .. try and hide various branches of the scene tree in your game project till you can find the minimum to produce an artifact, then try and reproduce it in a simpler project.

Some things which are some of the possible culprits:
* the viewport use
* your effects / fades
* custom shaders

It is likely there is more than 1 bug, the filtering / precision / wrapping looks different to the extra geometry bug. Geometry being drawn in the wrong place is a common bug because of the multiple matrices that are used in godot pipeline. In your second video lines seem to be coming out of the top left screen corner, which is odd, unless you are using that for an effect? Also are you using any procedural geometry (i.e. you create the verts yourself)?

**Comment by EIREXE** [comment]
I have managed to make it happen on a test project, seems like the shader used is the culprit.

[BatchingArtifacting.zip](https://github.com/godotengine/godot/files/5529652/BatchingArtifacting.zip)

**Comment by lawnjelly** [comment]
When I run your example I don't seem to be seeing any regressions (Linux Mint 18.3, Intel HD graphics 630). If you can confirm you are seeing something different, what GPU etc are you using?

**Comment by lawnjelly** [comment]
Getting there on this, I had to go out this afternoon, but the problem is occurring only when modulate is applied to the control (that has the style) that is non 1, 1, 1, 1 color. There is a different path when using white, because it affects baking colors. I suspect it is changing to large_fvf and this path may not have been fully tested with polys yet.

The style seems to be working by drawing a huge number of polys in the background to the control. Surprisingly large numbers, like 1900 polys for just one control, so this would have been really slow without batching.

**Comment by avencherus** [comment]
@lawnjelly Maybe not necessary, but here's what I see from my end with your revised MRP.

Win10 GeForce GTX 1660 SUPER/PCIe/SSE2 (GLES2 and GLES3)

![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/13004169/99045359-e230b900-2599-11eb-8f88-dba85522fb5c.gif)

**Comment by lawnjelly** [comment]
Yup it should be the same for everyone, it is not a GPU problem. It's just using the wrong shader for the vertex format because I hadn't implemented the custom shader stuff for polys. I'm nearly done sorting it now though. :+1: :grin:

Is only taking longer because I want to batch it properly rather than just revert to legacy.

**Comment by lawnjelly** [comment]
The PR above is now in for this poly corruption issue, and solves a bunch of other things too which haven't been noticed yet. I haven't tested yet but I don't think it will solve the texture filtering issue in @EIREXE project, as that is separate problem I think.

While testing it I also uncovered a regression in the 2d software skinning, unrelated. It is possible my fix for the samurai sword project #42576 could have broken something in the other skinning case. I'll try and get this fixed next.

**Comment by EIREXE** [comment]
Hello again, not sure if this is an issue on my end, but disabling the node TimingArm here makes batching work for the yellow hearts, however enabling it breaks it, is this because the notes all share the same texture as their source (using an atlas) and the timing arms do not? I also tried changing the reordering and lookahead values and that doesn't seem to do anything.

![imagen](https://user-images.githubusercontent.com/3101690/99527558-44742a00-299d-11eb-8f76-561bdda02bea.png)

![imagen](https://user-images.githubusercontent.com/3101690/99527633-68377000-299d-11eb-9907-350b36fc5883.png)

**Comment by lawnjelly** [comment]
> Hello again, not sure if this is an issue on my end, but disabling the node TimingArm here makes batching work for the yellow hearts, however enabling it breaks it, is this because the notes all share the same texture as their source (using an atlas) and the timing arms do not? I also tried changing the reordering and lookahead values and that doesn't seem to do anything.

Just to be clear, are you asking about a visual regression, or asking how to improve batching performance in this particular case? (I'm guessing you asking about improving batching).

To see what is happening the key is to use the `diagnose` mode in the project settings. This will log a list of commands and items every approx 10 seconds to the IDE.

Things that will typically break batching include texture and material changes. Item re-ordering can help with this type of situation, but not in all cases, and it will be limited by the reordering lookahead project setting.

I would probably need an example project to diagnose properly, but a diagnose log may help.

**Comment by EIREXE** [comment]
I see, that's the solution I expected then, thank you very much!

**Comment by oeleo1** [comment]
I tried 3.2.4-beta3. Great job! Most bugs I saw originally seem to have been addressed. I still have one visual regression which I would like to report here. It's a bit of a convoluted setup, but I believe you will grasp it easily.

This occurs on iOS target only (not on Android, MacOS, nor on Windows) and only in GLES2. Batching is disabled. Here is the minimal project I ended up with 
[sprite_test.zip](https://github.com/godotengine/godot/files/5632223/sprite_test.zip)

This basically is a Sprite with a custom shader swapping between 4 textures. There is a Light2D which is supposed to hide some parts of the screen containing the texture mask.  The problem is that what we see the Light2D texture in the Sprite, instead of the 1st sprite texture (out of the 4).

It sould look like this: 
![good](https://user-images.githubusercontent.com/40200623/100935292-33283300-34f0-11eb-8a2f-d3e8e889200b.gif)

But it looks like this:
![bad](https://user-images.githubusercontent.com/40200623/100935329-3d4a3180-34f0-11eb-8c48-3427e863ea75.gif)

Removing the light, shows the sprite correctly. GLES3 doesn't exhibit the problem.

NB: Not sure if this is related or not, but I notice that exporting to iOS requires me to check PVRTC compression in the settings.  This is wrong, but I will raise a bug about it separately. The project is common to different platforms, it used to work fine in 3.2.3 / iOS without that requirement, so this new requirement is not appropriate.

**Comment by oeleo1** [comment]
PS: I seem to remember seeing a commit by @clayjohn related to the number of texture slots or similar, but I just can't tell exactly what it was about. Could it be related? Very strange this occurs on iOS only and none of the other targets we usually work with.

**Comment by zzz-assault** [comment]
Godot version: 3.2.4 (Beta 3)

OS: Windows 10 
CPU: I7 8700
GPU: RTX 2080

Project used for testing is heavily using tilemaps to create a huge map (generated completely at the beginning, just optimized by visibility notifiers) - in the test case map size was 2k x 2k tiles -> both measurements are not 100% comparable due to proc gen content, but after several tests the result is basically not really affected.

I also haven't found any graphic issues so far! 

Comparison - stable 3.2.3 (GLES3) vs. beta 3 3.2.4 (GLES3):

![image](https://user-images.githubusercontent.com/47723585/100943964-8d2ff500-34fe-11eb-98e3-2a7bee35fdf3.png)

Btw. with activated Y-Sort the draw calls are still optimized, but the FPS identical (no improvement) -> I assume in this case bottleneck is Y-Sort logic instead of draw calls.

GREAT JOB lawnjelly!

**Comment by lawnjelly** [comment]
> This occurs on iOS target only (not on Android, MacOS, nor on Windows) and only in GLES2. Batching is disabled. Here is the minimal project I ended up with

If this occurs with batching off it sounds less likely to be a batching bug (so would be better in a separate issue, but I guess we can address here), however it may be a regression introduced by other changes to the renderer. You describe it as a regression, which previous version can you confirm it worked in?

You mention 3.2.3 but I'm not sure whether that is talking about PVRTC or the main problem you describe.

It is possible it is to do with @clayjohn's changes, but also highly likely is that you are simply running out of texture slots on your hardware. Mobile often has a maximum of 8 slots, and many of those are reserved in godot. This may be one for clayjohn in fact.

If you are running out of texture slots two things spring to mind as most obvious culprits:
* that this isn't a regression, it didn't work previously
* that the changes to the texture slots in clayjohn's PR has affected this, perhaps by making you reach the limit

Also, I see you are using `nvidia_workaround`. Have you tried it on the hardware both with and without the workaround? That is another possible.

**Comment by lawnjelly** [comment]
> Btw. with activated Y-Sort the draw calls are still optimized, but the FPS stay identical with and without batching (no improvement) -> I assume in this case bottleneck is Y-Sort logic instead of draw calls.

This does sound likely, if the drawcalls are reducing and the FPS isn't increasing it does usually mean the bottleneck is elsewhere.

**Comment by oeleo1** [comment]
> You describe it as a regression, which previous version can you confirm it worked in?

It is a regression from 3.2.3-stable for sure. I just went through the multicolored IDE of beta1 and can confirm that it works fine on the target as well. So this broke since beta2 or 3.

Sorry for reporting this issue here if it is unrelated to batching, but this is mainly due to my ignorance (and I'd be happy to open a separate issue in the bug tracker). I thought you were interested in any visual regressions compared to the legacy. Also, I _do_ want to test this much improved version of batching but need a non-batching performance basis first. Was torn away from the colorful beta1, skipped beta2 as too many residual visual regressions were still in. Now with beta3 I have only 2 visual regressions with batching turned off.

1. The messed textures I reported above
2. Sliders which doesn't fill properly. I believe this is a corollary of the ninepatch bug which was fixed after beta3 came out, so I'll wait on this until the next beta to confirm.

Re: PVRTC, yes, it seems totally unrelated.

Re: texture slots -- primary suspect at this point methinks. Note that the sprite & light are under a CanvasLayer. If I move them above the CanvasLayer, the problem goes away. If set the visibility of the light to false, the problem goes away.

**Comment by oeleo1** [comment]
> I just went through the multicolored IDE of beta1 and can confirm that it works fine on the target as well. So this broke since beta2 or 3.

Booh! I shamelessly lied. This is broken since beta1. Previously did the test with the wrong renderer (GLES3 instead of GLES2) but hey, the colorful IDE is so much fun to navigate and switch renderers. ;-)

**Comment by lawnjelly** [comment]
Yeah with the other bugs, if they are not batching, then 1 issue per bug is much easier to deal with, so I would encourage you to report them separately. Worst case they might be duplicates.

And also given the iOS nature of the 2d light bug, it is probably worth separating this too as it may require some verbosity to solve... you could just copy paste the info from here? If a canvas layer is involved this is also relevant.

**Comment by clayjohn** [comment]
@oeleo1 for reference here is the PR I made https://github.com/godotengine/godot/pull/42538.

It leaves some devices with only two custom texture slots instead of 3.

**Comment by oeleo1** [comment]
@clayjohn Right, thanks. The PR suggests 3 slots have been a favor for low end devices, but actually this happens on all iOS devices, including latest models (iPhones and iPads). One could argue that Apple/iOS is low end :-) but I am not into this debate. If there is a way out with 2 slots by refactoring stuff, I'll happily embrace it, but I don't see it (yet)...

**Comment by clayjohn** [comment]
@oeleo1 Fair enough, it looks like all iOS devices are limited to the minimum of 8 texture slots. :/

**Comment by oeleo1** [comment]
@lawnjelly @clayjohn Now onto some real feedback on the topic. This batching effort is an exceptional piece of work!

I did some profiling with Apples's Instruments on target hardware (5 year old iPhone SE 1st gen - supposedly a weak model as of this writing) without and with batching on a real action game. Batching was enabled with the default settings. Everything is GLES2.

Without batching, the main game load results in 56-60 FPS with regular dips down to 45 FPS on graphic intensive situations. Some particularly heavy scenes with thousands of (CPU) particles drop down to 25-30 FPS.

With batching, the number of draw calls is reduced by anything between 20% and 50%. The main game load results in steady 59-60 FPS with the graphic intensive situations dropping down to 56 FPS. I didn't see a lower number during the main play. The heavy scenes with thousands of particles drop down to the same amount of 25-30 FPS which is no surprise since particles are out scope.

I rarely engage in superlatives but what you've done here deserves the community's utmost respect. Bravo!

**Comment by summertimejordi** [comment]
I'm getting a Visual Regression when attempting to create a NinePatchRect from gdscript with batching enabled on GLES3.

On Godot 3.2.3, **No Batching (GLES3)**
![image](https://user-images.githubusercontent.com/12436824/101976232-06fb6900-3bf8-11eb-8e51-a17e0bb8f408.png)
^ This is what it's supposed to look like

On Godot 3.2.4, **_With_ Batching (GLES3)**
![image](https://user-images.githubusercontent.com/12436824/101976242-1da1c000-3bf8-11eb-92f8-6a05b4ccb289.png)
^ Incorrect. The border of the box is not being scaled correctly.

Is there a fix for this?

**Comment by anthonyromano** [comment]
Hi

I have a regression with batching. I've attached two screenshots, both running on beta 4, one with batching on one with batching off. On the stable build, batching on or off makes no visual difference. There are no Light2Ds in these screenshots. As you can see, in beta4 (and beta3 also had this bug, I did not test beta 1 and 2) the enemies vision cone fades out when it approaches the player. The player has a screen reading shader around him to create the "light" effect. The enemy vision cones are also screen reading shaders. The issue seems to be when they overlap

Batching On (Beta4):
<img width="400" alt="Screen Shot 2020-12-12 at 1 55 03 AM" src="https://user-images.githubusercontent.com/5588784/101977707-8bacae00-3c1d-11eb-89ca-ad3862c87e84.png">

Batching Off (Beta4) (And what batching on and off look like normally)
<img width="407" alt="Screen Shot 2020-12-12 at 1 54 24 AM" src="https://user-images.githubusercontent.com/5588784/101977708-8bacae00-3c1d-11eb-9486-ab1c424bf231.png">

**Comment by lawnjelly** [comment]
> I'm getting a Visual Regression when attempting to create a NinePatchRect from gdscript with batching enabled on GLES3.

There are now 2 ninepatch modes, see `project_settings->rendering->quality->2d->ninepatch_mode`, you may be relying on behaviour from the puthre method (scaling).

**Comment by lawnjelly** [comment]
> I have a regression with batching. I've attached two screenshots, both running on beta 4, one with batching on one with batching off. On the stable build, batching on or off makes no visual difference. There are no Light2Ds in these screenshots. As you can see, in beta4 (and beta3 also had this bug, I did not test beta 1 and 2) the enemies vision cone fades out when it approaches the player. The player has a screen reading shader around him to create the "light" effect. The enemy vision cones are also screen reading shaders. The issue seems to be when they overlap.

I'll take a look at this when I'm home, but I'll need a minimum reproduction project.

**Comment by summertimejordi** [comment]
> > I'm getting a Visual Regression when attempting to create a NinePatchRect from gdscript with batching enabled on GLES3.
> 
> There are now 2 ninepatch modes, see `project_settings->rendering->quality->2d->ninepatch_mode`, you may be relying on behaviour from the puthre method (scaling). Try the other setting, and let us know if it fixes for you.

With NinePatch Mode set to  **default**:
![Screenshot_20201212_110952](https://user-images.githubusercontent.com/12436824/101990208-3bf4d380-3c6b-11eb-8748-4adcf685c4d1.png)

With NinePatch Mode set to  **scaling**:
![Screenshot_20201212_111024](https://user-images.githubusercontent.com/12436824/101990213-41521e00-3c6b-11eb-9f78-c1c1090699fc.png)

It's still not correct compared to Godot 3.2.3:
![image](https://user-images.githubusercontent.com/12436824/101990228-60e94680-3c6b-11eb-8f4d-30455a072145.png)

**Comment by lawnjelly** [comment]
@summertimejordi If you can make a min reproduction project I can take a look. :+1:

**Comment by summertimejordi** [comment]
> @summertimejordi If you can make a min reproduction project I can take a look.

I just finished doing that when you replied :+1: 

Here it is:
https://drive.google.com/file/d/1xODicuJRJrrSoIInFa4IEugyAJbIW9o-/view?usp=sharing

**Comment by lawnjelly** [comment]
Ok I've done a little investigating. When I ran with `flash_batching` it shows that the legacy renderer is the same as the batching (in fact in your case it is using the legacy path). I had to remind myself of the changes to ninepatches as this was a couple of months ago.

Here is some relevant discussion:
https://github.com/godotengine/godot-proposals/issues/1618
https://github.com/godotengine/godot/pull/42119#issuecomment-704097907
https://github.com/godotengine/godot/pull/32170

TLDR - there was an old method that was used for ninepatches. Puthre made a PR which changed this behaviour, to the behaviour you were using in your example. This change proved controversial - there was a lot of pushback with people saying they preferred the old behaviour.

In unified batching, you now get the choice between using the old old method (default) and the old method (scaling). *However* due to technical reasons, you only get this choice when not using tiled mode for the ninepatch. The only way to support tiled mode is through the GLES3 shader (it is not supported in GLES2), and it was decided not to be possible / practical to support both methods in the shader. We had to choose one, and at the time @clayjohn and myself got the impression that the popular approach which was the old old approach.

However, all is not lost, in your case, if you use batching, and you change the mode to `scaling` (in the project setting) and change the tile mode to AXIS_STRETCH_MODE_STRETCH in both axes, then it can use batching and emulate the scaling method.

Where this won't work is if you have absolutely have to use AXIS_STRETCH_MODE_TILE_FIT, as this can only be done via shader. I don't think you have to in your test case (try it out).

This is just the consequence of us having to make a hard choice between one method and the other in the shader. There is no right answer, either way some people will be happy and some disappointed.

**Comment by summertimejordi** [comment]
> Ok I've done a little investigating. When I ran with `flash_batching` it shows that the legacy renderer is the same as the batching (in fact in your case it is using the legacy path). I had to remind myself of the changes to ninepatches as this was a couple of months ago.
> 
> Here is some relevant discussion:
> https://github.com/godotengine/godot-proposals/issues/1618
> https://github.com/godotengine/godot/pull/42119#issuecomment-704097907
> https://github.com/godotengine/godot/pull/32170
> 
> TLDR - there was an old method that was used for ninepatches. Puthre made a PR which changed this behaviour, to the behaviour you were using in your example. This change proved controversial - there was a lot of pushback with people saying they preferred the old behaviour.
> 
> In unified batching, you now get the choice between using the old old method (default) and the old method (scaling). *However* due to technical reasons, you only get this choice when not using tiled mode for the ninepatch. The only way to support tiled mode is through the GLES3 shader (it is not supported in GLES2), and it was decided not to be possible / practical to support both methods in the shader. We had to choose one, and at the time @clayjohn and myself got the impression that the popular approach which was the old old approach.
> 
> However, all is not lost, in your case, if you use batching, and you change the mode to `scaling` (in the project setting) and change the tile mode to AXIS_STRETCH_MODE_STRETCH in both axes, then it can use batching and emulate the scaling method. If you don't have to use the tile stretch mode, not only will you get the choice of how you look, but it should render much faster.
> 
> Where this won't work is if you have absolutely have to use AXIS_STRETCH_MODE_TILE_FIT, as this can only be done via shader. I don't think you have to in your test case (try it out).
> 
> This is just the consequence of us having to make a hard choice between one method and the other in the shader. There is no right answer, either way some people will be happy and some disappointed.
> 
> BTW here's a screencap in 3.2.4 using the different stretch modes, and ninepatch mode set to scaling:
> ![ninepatch](https://user-images.githubusercontent.com/21999379/101993097-c3fccc80-3caf-11eb-9e02-3e166f10a40b.gif)
> 


Oh, I didn't know that. Yeah in my case I would be totally fine with using Stretch, thanks!

**Comment by oeleo1** [comment]
Not sure I understand the explanation fully but to followup on this, I also have a ninepatch related regression with a slider as follows:

3.2.3-stable (good)
![slider_good](https://user-images.githubusercontent.com/40200623/101994430-a84ef180-3cc2-11eb-954b-04b81c624111.png)

3.2.4-beta 3/4 (bad)
![slider_bad](https://user-images.githubusercontent.com/40200623/101994438-b7ce3a80-3cc2-11eb-955f-3387a2cf458e.png)

The bad shows both in the editor and at runtime. Changing the `Rendering -> Quality -> Ninepatch Mode` from `Default` to `Scaling` shows it right at runtime. Minimal project: 
[slider_test.zip](https://github.com/godotengine/godot/files/5683728/slider_test.zip)

Bug or feature ?

**Comment by lawnjelly** [comment]
> The bad shows both in the editor and at runtime. Changing the `Rendering -> Quality -> Ninepatch Mode` from `Default` to `Scaling` shows it right at runtime. Minimal project:
> [slider_test.zip](https://github.com/godotengine/godot/files/5683728/slider_test.zip)
> 
> Bug or feature ?

I'm not quite sure yet, I'm not familiar with these UI items. As far as I can tell, it is drawing a slider progress (which is the coloured background), and an HSlider, which includes the godot button AND I'm presuming a ninepatch, which is the black bar. It is set to StyleBoxEmpty for the HSlider, which the documentation suggests should not be drawing anything, but it *is* drawing something. It could be relying on some behaviour of the old shader to draw the ninepatch but have nothing appear on screen.

I'll see if I can debug into the ninepatch call.

**Comment by oeleo1** [comment]
Too low-level to engage in a meaningful discussion, but please note that the TextureProgress has a `nine_patch_stretch` property, set to on, with all stretch margins set to 0. I want it that way, otherwise it doesn't fill the rectangle when it is set to off (3.2.3 behavior). When it is off, it is transparent (black underneath).

With 3.2.4 - same settings, but when the `nine_patch_stretch` is switched off, one can see the gray color of the default slider theme.

What I want is the 3.2.3 behavior both in the editor and at runtime. If 3.2.4 changed things via the new `Settings` ninepatch scaling switch, I'd like t to show as such in the editor too.

**Comment by lawnjelly** [comment]
I really didn't understand your previous post, whether you were trying to report another bug or give more info on the first bug. However the bug you reported is fixed in the PR above.

**Comment by oeleo1** [comment]
> Sorry I really didn't understand your previous post

The intent was to provide more info but I may have unwillingly reported another bug :-) In any case, many thanks for your prompt response and related fix!

**Comment by EIREXE** [comment]
![imagen](https://user-images.githubusercontent.com/3101690/107140493-9edcb100-6922-11eb-8b85-e0ecaa8e6757.png)

Discovered what I believe is a new crash, I'm going to try to replicate it without multithreaded rendering enabled to see if it's related.

**Comment by lawnjelly** [comment]
> Discovered what I believe is a new crash, I'm going to try to replicate it without multithreaded rendering enabled to see if it's related.

Ah that's very useful the screenshot, let's me know roughly what is going wrong. It is probably some combination of item types that are being put in the same joined item incorrectly, or maybe this is occurring because of a custom draw. Are you using custom drawing? And which primitives are you using in your game (i.e. sprite, poly, line, multimesh2d etc)?

You may be able to get it more likely to hit this problem by settings `rendering->batching->parameters->batch_buffer_size` to the smallest value. It is basically running out of vertices in the temporary buffer, which shouldn't happen, because logic earlier on should have split these apart. So that logic earlier is where the bug is.

Let me know if you can create a min reproduction project or find anything else that might help.

**Comment by EIREXE** [comment]
> 
> 
> > Discovered what I believe is a new crash, I'm going to try to replicate it without multithreaded rendering enabled to see if it's related.
> 
> Ah that's very useful the screenshot, let's me know roughly what is going wrong. It is probably some combination of item types that are being put in the same joined item incorrectly, or maybe this is occurring because of a custom draw. Are you using custom drawing? And which primitives are you using in your game (i.e. sprite, poly, line, multimesh2d etc)?
> 
> You may be able to get it more likely to hit this problem by settings `rendering->batching->parameters->batch_buffer_size` to the smallest value. It is basically running out of vertices in the temporary buffer, which shouldn't happen, because logic earlier on should have split these apart. So that logic earlier is where the bug is.
> 
> Let me know if you can create a min reproduction project or find anything else that might help.
> 
> It may be related to this issue reported on reddit which sounds similar:
> https://www.reddit.com/r/godot/comments/l6wa9y/release_candidate_godot_324_rc_1_nows_a_great/glfn98p?utm_source=share&utm_medium=web2x&context=3

What constitutes as custom drawing? I believe the crash happens randomly when showing a scene that I believe doesn't use any custom drawing, but the scene before it (that gets removed from the tree but not freed) does use custom drawing in the form of a shader material that is inherited by children.

![imagen](https://user-images.githubusercontent.com/3101690/107147550-f42eb780-694e-11eb-88d4-0513ff34ab8a.png)

I will try to rip this scene out of the game and see if I can make it happen outside of it.

**Comment by lawnjelly** [comment]
Custom drawing is where you create your own `_draw` function in a 2d object, and tell it to draw e.g. polys, lines, rects manually.

**Comment by EIREXE** [comment]
> 
> 
> Custom drawing is where you create your own `_draw` function in a 2d object, and tell it to draw e.g. polys, lines, rects manually.

Indeed in this case no custom _draw functions are being used, only built-in nodes.

**Comment by Reapetitive** [comment]
The "max join items" of batching feature seems to make some seemingly random textures flicker and change their modulation incorrectly if near lights. Setting it to 0 doesnt hit the performance for me and fixes the problem.

**Comment by lawnjelly** [comment]
> The "max join items" of batching feature seems to make some seemingly random textures flicker and change their modulation incorrectly if near lights. Setting it to 0 doesnt hit the performance for me and fixes the problem.

With a minimum reproduction project, I can investigate.

**Comment by djrain** [comment]
Is looking at Performance.RENDER_2D_DRAW_CALLS_IN_FRAME a valid way to eyeball how well batching is working? I'm seeing some mixed results from that, which don't always seem to make sense.

**Comment by lawnjelly** [comment]
Yes, but I would encourage reading the docs (we have to update this to say GLES3 as well, but essentially it's the same):
https://docs.godotengine.org/en/3.2/tutorials/optimization/batching.html

The relationship between your scene and how well it gets batched is _complicated_. If you are using features that can't be batched, or arrangements, you'll get less batching.

**Comment by syntax-errol** [comment]
![scrot-2021-03-11-11:31:26](https://user-images.githubusercontent.com/80454540/110768647-b5469700-8257-11eb-8cf0-d5afa1fc29d1.png)
I am getting black lines between my tilemap-quadrants when using scissor_area_threshold. It can vary in width (0-2 pixels). turning scissor_area_threshold to 1 fixes the issue. I have tried to tweak other options but without change.
The effect can appear when more lights are added near the tile quadrants.

(10 y.o. computer with low end nvidia card on linux)

**Comment by lawnjelly** [comment]
Ah I have a vague memory of a report like this before. It could be an off by one error in the clip rectangle. I will take a look, it should be easy to fix. If you have a reproduction project that would be helpful. :+1:

**Comment by syntax-errol** [comment]
I created a test project. you can find it here: http://orbmit.com/godot/line_error_test.zip

![scrot-2021-03-11-12:29:01](https://user-images.githubusercontent.com/80454540/110773867-7c112580-825d-11eb-9968-5f75946696c9.png)

**Comment by lawnjelly** [comment]
Fantastic. I've just made a PR but this will be good to test it. :+1:

**Comment by bones527** [comment]
The debug collision shapes probably aren't batched.When debugging with a lot of collision shapes visible it slows down a bit and the amount of draw calls increase a lot. Not a huge problem. Just letting you know.

**Comment by djrain** [comment]
Doing some testing, and I am unable to reproduce the scenario described in the documentation - 2 nonoverlapping sprites with same texture, and 3 lights affecting these. The docs say this should be done in 4 draw calls, but it always shows 8 for me.

The debug prints this:

canvas_begin FRAME 582
items
	joined_item 1 refs, hxform
			batch R 0-1 [0 - 152] {255 255 255 255 }
			batch R 0-1 [0 - 152] {255 255 255 255 }
			batch R 0-1 [0 - 152] {255 255 255 255 }
			batch R 0-1 [0 - 152] {255 255 255 255 }
	joined_item 1 refs, hxform
			batch R 0-1 [0 - 152] {255 255 255 255 }
			batch R 0-1 [0 - 152] {255 255 255 255 }
			batch R 0-1 [0 - 152] {255 255 255 255 }
			batch R 0-1 [0 - 152] {255 255 255 255 }
canvas_end
	light items joined: 1

I'm still somewhat confused on exactly how to interpret this, but they are not getting batched, correct? Is this odd?
Batching settings are default, and I'm using 3.2.4 RC5 with GLES3.

**Comment by lawnjelly** [comment]
If you provide a test project I can take a look, it may be a bug, or it may be intended behaviour.

**Comment by djrain** [comment]
I'm afraid there are problems with normal maps in 3.3 RC6 :(

 I made an issue: #47276

**Comment by oeleo1** [comment]
Another regression with rc6 [here](https://github.com/godotengine/godot/issues/47285). Sorry lawnjelly.

**Comment by djrain** [comment]
So I'm experiencing a nasty issue with batching and lights, where it seems like the light gets cut off around the right and bottom edges of the screen somewhat randomly...

https://user-images.githubusercontent.com/33777501/119877369-ce1d2b00-bedd-11eb-9957-4bdf5d5a0803.mov

It happens both in game and in editor, but only when batching is on AND scissor area threshold is less than 1. Also worth noting that it only seems to affect my terrain that is custom drawn using draw_primitive(), it doesn't affect sprites. As you can see, just changing zoom in editor can make it appear, it is also "fixed" when I just draw some extra things in draw function. Very odd. I think this must be a bug? Haven't managed to reproduce it in a small project yet.

It seems to happen in all versions from 3.3.2 going back to 3.2.3, and affects GLES2 as well.

**Comment by lawnjelly** [comment]
> So I'm experiencing a nasty issue with batching and lights, where the light gets cut off on a certain node in a certain area of the screen somewhat randomly...

That's interesting. It's just possible the AABB is incorrect for a custom draw. But I'm guessing, will have to wait for a minimum reproduction project.

**Comment by djrain** [comment]
Finally managed to narrow it down, but turns out it's not a batching issue specifically. Anyhow, issue with MRP here: #49156

**Comment by djrain** [comment]
Is there a reason why draw_primitive() calls can't be batched? Identical calls to draw_polygon() get batched, but not draw_primitive(). I can see why this might be the case, since it handles different primitive types... if so, that should probably be documented.

**Comment by akien-mga** [comment]
I'll close this issue as it has served its purpose as a catch-all for testing results prior to the 3.3 release.

For new issues with batching in currently supported Godot versions, please open a dedicated bug report so that we can track each issue individually.

If you came here to share not a bug but a report of amazing performance improvement in your project, feel free to share it here anyway ;)
